### ISSUE TITLE
torch pruning + DETR 

### ISSUE THREAD

**Issue by Faisal-Hajari** [VainF/Torch-Pruning#307]
when using the code with DETR models I find it incompatible. I think the tracing doesn't work well with DETR. Has anyone used it with any implementation of DETR ? 

In my case. If I restrict the pruning to the backbone of the model it works fine. but when I pass the entire model is when I start facing issues.

**Comment by ganliqiang** [comment]
how do you prune the part of a model. can you share some codes? when i prune the parf od a model,the error occurs: no forword in your model?

**Comment by jn12-29** [comment]
> how do you prune the part of a model. can you share some codes? when i prune the parf od a model,the error occurs: no forword in your model?

You can just set other layers in the ignored_layers of pruner.

**Comment by jn12-29** [comment]
from transformers import DetrForObjectDetection
import torch
import torch_pruning as tp

pruning_ratio = 0.25

device = "cuda" if torch.cuda.is_available() else "cpu"

model = DetrForObjectDetection.from_pretrained("./detr-resnet-50").to(device)

print(model)
# replace the DetrFrozenBatchNorm2d with nn.BatchNorm2d
name_list = []
new_module_list = []
module_name_dict = {name: module for name, module in model.model.named_modules()}
from transformers.models.detr.modeling_detr import DetrFrozenBatchNorm2d
for name, module in module_name_dict.items():
    if isinstance(module, DetrFrozenBatchNorm2d):
        ind = name.rfind(".")
        if ind == -1:
            father = module_name_dict[""]
        else:
            father = module_name_dict[name[:ind]]
        weight, bias, running_mean, running_var = module.weight, module.bias, module.running_mean, module.running_var
        new_module = torch.nn.BatchNorm2d(num_features=weight.shape[0])
        new_module.weight.data.copy_(weight)
        new_module.bias.data.copy_(bias)
        new_module.running_mean.data.copy_(running_mean)
        new_module.running_var.data.copy_(running_var)
        setattr(father, name[ind+1:], new_module)
        print(f'replace {name} with {type(new_module)}')
model.to(device)

print(model)

model.train()
for p in model.model.backbone.conv_encoder.parameters():
    p.requires_grad = True
    
example_inputs = torch.randn(1, 3, 224, 224).to(device)

base_macs, base_nparams = tp.utils.count_ops_and_params(model, example_inputs)


importance = tp.importance.GroupNormImportance(p=2, group_reduction='mean')
ignored_layers = [model.class_labels_classifier, model.bbox_predictor.layers[-1]]
# ignored_layers += [model.model.input_projection, model.model.query_position_embeddings, model.model.encoder, model.model.decoder]
pruner = tp.pruner.MetaPruner(
    model,
    example_inputs,
    importance=importance,
    pruning_ratio=pruning_ratio,
    ignored_layers=ignored_layers
)
DG = tp.DependencyGraph().build_dependency(model, example_inputs)

for g in pruner.step(interactive=True):
    # len(g._group) == 193 is the decoder layers, as g will not prune v_proj in the first layer of decoder, we prune it manually
    if len(g._group) == 193:
        _, idxs = g._group[-1]
        group = DG.get_pruning_group(model.model.decoder.layers[0].self_attn.v_proj, tp.prune_linear_in_channels, idxs=idxs)
        group.prune()
    g.prune()

# update the position_embedding dim
from transformers.models.detr.modeling_detr import DetrSinePositionEmbedding
old_pos_embed = model.model.backbone.position_embedding
model.model.backbone.position_embedding = DetrSinePositionEmbedding(int(old_pos_embed.embedding_dim * (1 - pruning_ratio)), normalize=True)

# update the query_position_embeddings dim
from torch.nn import Embedding
old_qpe = model.model.query_position_embeddings
model.model.query_position_embeddings = Embedding(num_embeddings=old_qpe.num_embeddings, embedding_dim=int(old_qpe.embedding_dim * (1 - pruning_ratio)), padding_idx=old_qpe.padding_idx, max_norm=old_qpe.max_norm, norm_type=old_qpe.norm_type, scale_grad_by_freq=old_qpe.scale_grad_by_freq, sparse=old_qpe.sparse)

# update the transformer atten dim
from transformers.models.detr.modeling_detr import DetrAttention
for module in model.model.modules():
    if isinstance(module, DetrAttention):
        module.embed_dim = int(module.embed_dim * (1 - pruning_ratio))
        module.head_dim = int(module.head_dim * (1 - pruning_ratio))


model.to(device)
print(model)
pruned_macs, pruned_nparams = tp.utils.count_ops_and_params(model, example_inputs)
# print(model)
print(f'\nPruning ratio: {pruning_ratio}')
print("Before Pruning: MACs=%f G, #Params=%f G"%(base_macs/1e9, base_nparams/1e9))
print("After Pruning: MACs=%f G, #Params=%f G"%(pruned_macs/1e9, pruned_nparams/1e9))

model.to('cpu')
for p in model.parameters():
    p.requires_grad = False
torch.save(model, 'detr-resnet-50-pruned-0.25.pth')

An example that has worked for me.
